File: 1026. Maximum Difference Between Node and Ancestor.py
'''Given the root of a binary tree, find the maximum value v for which there exist different nodes a and b where v = |a.val - b.val| and a is an ancestor of b.

A node a is an ancestor of b if either: any child of a is equal to b or any child of a is an ancestor of b.'''

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    cache = {}
    def maxChild(self, root):
        if root==None:
            return 0
        if (root.left, "max") in self.cache:
            left = self.cache[(root.left, "max")]
        else:
            left = self.maxChild(root.left)
            self.cache[(root.left, "max")] = left

        if (root.right, "max") in self.cache:
            right = self.cache[(root.right, "max")]
        else:
            right = self.maxChild(root.right)
            self.cache[(root.right, "max")] = right
        return max(root.val,left,right)

    def minChild(self, root):
        if root==None:
            return 10**6
        if (root.left, "min") in self.cache:
            left = self.cache[(root.left, "min")]
        else:
            left = self.minChild(root.left)
            self.cache[(root.left, "min")] = left

        if (root.right, "min") in self.cache:
            right = self.cache[(root.right, "min")]
        else:
            right = self.minChild(root.right)
            self.cache[(root.right, "min")] = right
        return min(root.val,left,right)

    def maxAncestorDiff(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root==None:
            return 0
        l,r = 0,0
        sl,lr = 0,0

        if root.left:
            if (root.left, 's') in self.cache:
                sl = self.cache[(root.left, 's')]
            else:
                sl = self.minChild(root.left)
                self.cache[(root.left, 's')] = sl

            if (root.left, 'l') in self.cache:
                ll = self.cache[(root.left, 'l')]
            else:
                ll = self.maxChild(root.left)
                self.cache[(root.left, 'l')] = ll
            l = max(abs(root.val - sl), abs(root.val - ll))

        if root.right:
            if (root.right, 'l') in self.cache:
                lr = self.cache[(root.right, 'l')]
            else:             
                lr = self.maxChild(root.right)
                self.cache[(root.right, 'l')] = lr

            if (root.right, 's') in self.cache:
                sr = self.cache[(root.right, 'l')]
            else:
                sr = self.minChild(root.right) 
                self.cache[(root.right, 'l')] = sr
            r =  max(abs(root.val - lr), abs(root.val - sr))

        n = max(l,r)
        al,ar = 0,0
        if root.left and root.left in self.cache:
            al = self.cache[root.left]
        elif root.left:
            al = self.maxAncestorDiff(root.left)
            self.cache[root.left] = al

        if root.right and root.right in self.cache:
            ar = self.cache[root.right]
        elif root.right:           
            ar = self.maxAncestorDiff(root.right)
            self.cache[root.right] = ar

        return max(n, al, ar)

        

File: 1026. Maximum Difference Between Node and Ancestor.c
/*
Given the root of a binary tree, find the maximum value v for which there exist different nodes a and b where v = |a.val - b.val| and a is an ancestor of b.

A node a is an ancestor of b if either: any child of a is equal to b or any child of a is an ancestor of b.
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

 // compare current with the max and min of the currents ancestors/subtree nodes

# define INF 1000000
int min(int a, int b){
    return a<b?a:b;
}
int max(int a, int b){
    return a>b?a:b;
}
int min_node(struct TreeNode* root){
    if(root == NULL){
        return INF;
    }
    return min(root->val, min(min_node(root->left), min_node(root->right)));
}
int max_node(struct TreeNode* root){
    if(root == NULL){
        return 0;
    }
    return max(root->val, max(max_node(root->left), max_node(root->right)));
}

int maxAncestorDiff(struct TreeNode* root) {
    if(root == NULL){
        return 0;
    }
    int left_min = root->val;
    int left_max = root->val;
    if(root->left){
        left_min = min_node(root->left);
        left_max = max_node(root->left);
    }
    int right_min = root->val;
    int right_max = root->val;
    if(root->right){
        right_min = min_node(root->right);
        right_max = max_node(root->right);
    }
    int smallest = min(left_min, right_min);
    int largest = max(left_max, right_max);
    int diff = max(abs(smallest-root->val), abs(largest-root->val));

    return max(diff, max(maxAncestorDiff(root->left), maxAncestorDiff(root->right)));    
}